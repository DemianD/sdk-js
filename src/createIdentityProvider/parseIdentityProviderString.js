import { createIdentityProviderError } from '../util/createError'

import { MISCONFIGURED, MISCONFIGURED_AUTH_PROVIDER } from '../constants'

export default function parseIdentityProviderString(idp_descriptor = '') {
  const [type, encoded_options] = idp_descriptor.split('://')

  if (type === undefined || type === '') {
    throw createIdentityProviderError(
      'You should provide a `type` as identity provider.',
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  if (encoded_options === undefined || encoded_options === '') {
    throw createIdentityProviderError(
      'You should provide `options` for the identity provider.',
      MISCONFIGURED,
      MISCONFIGURED_AUTH_PROVIDER
    )
  }

  const options = encoded_options
    .split('&')
    .map(item => item.split('='))
    .map(([key, value]) => ({
      [key]: decodeURIComponent(value),
    }))
    .reduce((a, b) => Object.assign(a, b))

  return { type, options }
}
