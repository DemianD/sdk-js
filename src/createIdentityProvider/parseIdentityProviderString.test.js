import parseIdentityProviderString from './parseIdentityProviderString'

describe('parseIdentityProviderString', () => {
  it('should be possible to parse the string', () => {
    expect(parseIdentityProviderString('type://foo=bar&bar=baz')).toEqual({
      type: 'type',
      options: {
        foo: 'bar',
        bar: 'baz',
      },
    })
  })

  it('should not throw an error when there are no options', () => {
    expect(() => parseIdentityProviderString('type://')).not.toThrowError()
  })

  it('should throw an error when there is no type but there are options', () => {
    expect(() =>
      parseIdentityProviderString('://foo=bar&bar=baz')
    ).toThrowErrorMatchingSnapshot()
  })

  it('should throw an error when there is no type and no options', () => {
    expect(() =>
      parseIdentityProviderString('://')
    ).toThrowErrorMatchingSnapshot()
  })

  it('should throw an error when there is an invalid value', () => {
    expect(() => parseIdentityProviderString('')).toThrowErrorMatchingSnapshot()
    expect(() => parseIdentityProviderString()).toThrowErrorMatchingSnapshot()
    expect(() =>
      parseIdentityProviderString(null)
    ).toThrowErrorMatchingSnapshot()
  })
})
