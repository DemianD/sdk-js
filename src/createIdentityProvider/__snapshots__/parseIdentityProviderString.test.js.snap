// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseIdentityProviderString should throw an error when there is an invalid value 1`] = `"[IDENTITY PROVIDER]: You should provide a \`type\` as identity provider."`;

exports[`parseIdentityProviderString should throw an error when there is an invalid value 2`] = `"[IDENTITY PROVIDER]: You should provide a \`type\` as identity provider."`;

exports[`parseIdentityProviderString should throw an error when there is an invalid value 3`] = `"Cannot read property 'split' of null"`;

exports[`parseIdentityProviderString should throw an error when there is no type and no options 1`] = `"[IDENTITY PROVIDER]: You should provide a \`type\` as identity provider."`;

exports[`parseIdentityProviderString should throw an error when there is no type but there are options 1`] = `"[IDENTITY PROVIDER]: You should provide a \`type\` as identity provider."`;
